numpy/random/mtrand/mtrand.pyx

```diff
@@ -574,6 +579,128 @@ def _shape_from_size(size, d):
            shape = tuple(size) + (d,)
     return shape
 
+
+# Set up dictionary of integer types and relevant functions.
+#
+# The dictionary is keyed by dtype(...).name and the values
+# are a tuple (low, high, function), where low and high are
+# the bounds of the largest half open interval `[low, high)`
+# and the function is the relevant function to call for
+# that precision.
+
+
+def rand_int32(low, high, size, rngstate):
+    """
+    random_int32(self, low, high=None, size=None, state)
+
+    Return random np.int32 integers between `low` and `high`, inclusive.
+
+    Return random integers from the "discrete uniform" distribution in the
+    closed interval [`low`, `high`).  If `high` is None (the default),
+    then results are from [0, `low`).
+
+    Parameters
+    ----------
+    low : int
+        Lowest (signed) integer to be drawn from the distribution (unless
+        ``high=None``, in which case this parameter is the *highest* such
+        integer).
+    high : int
+        If provided, the largest (signed) integer to be drawn from the
+        distribution (see above for behavior if ``high=None``).
+    size : int or tuple of ints
+        Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
+        ``m * n * k`` samples are drawn.  Default is None, in which case a
+        single value is returned.
+
+    Returns
+    -------
+    out : int or ndarray of npy_intp
+          `size`-shaped array of random integers from the appropriate
+          distribution, or a single such random int if `size` not provided.
+
+    """
+    cdef npy_uint32 off, rng, buf
+    cdef npy_uint32 *out
+    cdef ndarray array "arrayObject"
+    cdef npy_intp cnt
+    cdef rk_state *state = <rk_state *>NpyCapsule_AsVoidPtr(rngstate)
+
+    rng = <npy_uint32>(high - low)
+    off = <npy_uint32>(<npy_int32>low)
+    if size is None:
+        rk_random_uint32(off, rng, 1, &buf, state)
+        return <npy_int32>buf
+    else:
+        array = <ndarray>np.empty(size, np.int32)
+        cnt = PyArray_SIZE(array)
+        out = <npy_uint32 *>PyArray_DATA(array)
+        with nogil:
+            rk_random_uint32(off, rng, cnt, out, state)
+        return array
```

Comment:
At the bit level, there is no difference between signed/unsigned addition and subtraction on two's complement machines. However, unsigned arithmetic is specified to be modular, so overflow is always ignored. For signed integers the behavior may be undefined. Here we need to be careful to convert signed integers to two's complement signed binary integers and then treat those integers as unsigned, hence the double cast in some cases. The difference needs to be a positive so the generated random offsets can the added to the offset. To see how this works in practice, think of the unsigned numbers a located on a circle, increasing clockwise, with 0 just after the largest, and call 0 the wrapping point. Signed arithmetic is on the same circle, but the wrapping point is chosen to be at the most negative value represented by `0x100....0` on the circle. Addition of positive numbers moves a point clockwise around the circle in both cases. To make sure every thing works as expected the low and high bounds are checked before the call so that proceeding clockwise from the lowest to the highest never crosses the chosen wrapping point.

Note that another option for representing signed numbers is offset binary, which can be advantageous in some circuit designs and DSP.